### PR TITLE
Instant Action v2

### PR DESCRIPTION
**Pull request by ConstantinKV**
gitlab issue 170

### CODE REVIEW

File: app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt [hunk 1]
```diff
@@ -163,6 +170,50 @@ class SubmitActionViewModel(
...
+    private fun showConnectionsSelector(connections: List<Connection>) {
+        val result = connections.convertConnectionsToViewModels(
+            context = appContext,
+            locationManager = locationManager
+        )
+        viewMode = ViewMode.SELECT
+        showConnectionsSelectorFragmentEvent.postValue(ViewModelEvent(dataBundle(result)))
+    }
+
+    private fun openReturnToUrl() {
```
Comment (mnewlive) [inline app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt:183]: it is not clear what the function does, judging by the name of the event, it opens a link.

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/list/ConnectionsListInteractor.kt [hunk 2]
```diff
@@ -98,7 +98,7 @@ class ConnectionsListInteractor(
         if (connection.apiVersion == API_V1_VERSION) {
             apiManagerV1.revokeConnections(connectionsAndKeys = listOf(richConnection), resultCallback = null)
         } else if (connection.apiVersion == API_V2_VERSION) {
-            apiManagerV2.revokeConnections(connections = listOf(richConnection), callback = null)
+            apiManagerV2.requestRevokeConnections(connections = listOf(richConnection), callback = null)
```
Comment (mnewlive) [inline app/src/main/kotlin/com/saltedge/authenticator/features/connections/list/ConnectionsListInteractor.kt:101]: mb request To revoke connections?

Comment (mnewlive) [reply app/src/main/kotlin/com/saltedge/authenticator/features/connections/list/ConnectionsListInteractor.kt:101]: And I think you should choose one thing. Check please `apiManagerV1.revokeConnections`

Comment (mnewlive) [reply app/src/main/kotlin/com/saltedge/authenticator/features/connections/list/ConnectionsListInteractor.kt:101]: pull down the current changes

File: app/src/main/kotlin/com/saltedge/authenticator/features/main/MainActivityInteractor.kt [hunk 3]
```diff
@@ -47,7 +47,7 @@ class MainActivityInteractor(
             .mapNotNull { it.toRichConnection(keyStoreManager) }
         apiManagerV1.revokeConnections(connectionsAndKeys = richConnections
```
Comment (mnewlive) [inline app/src/main/kotlin/com/saltedge/authenticator/features/main/MainActivityInteractor.kt:48]: Check please comments about the same naming

File: authenticator_core/src/main/java/com/saltedge/authenticator/core/model/ActionAppLinkData.kt [hunk 4]
```diff
@@ -20,10 +20,36 @@
...
-    var connectUrl: String,
+    var apiVersion: String,
+    var actionIdentifier: ID,
+    var providerID: ID?,
+    var connectUrl: String?,
     var returnTo: String?
-) : Serializable
+) : Serializable {
+    val isV2Api: Boolean
+        get() = apiVersion == "2"
```
Comment (mnewlive) [inline authenticator_core/src/main/java/com/saltedge/authenticator/core/model/ActionAppLinkData.kt:36]: Use please constants

File: app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt [hunk 5]
```diff
@@ -70,90 +77,90 @@ class SubmitActionViewModel(
...
+            else -> showConnectionsSelector(connections)
         }
     }
 
-    fun showConnectionSelector(connectionGuid: GUID) {
-        this.richConnection = connectionsRepository.getByGuid(connectionGuid)?.toRichConnection(keyStoreManager)
-        viewMode = if (richConnection == null) ViewMode.ACTION_ERROR else ViewMode.START
-        onViewCreated()
+    fun onViewCreated() {
+        val currentRichConnection = richConnection
```
[the remark this answers is not in the document]
Comment (ConstantinKV) [reply app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt:94]: alternative of the `let`

File: app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt [hunk 6]
```diff
@@ -70,90 +77,90 @@ class SubmitActionViewModel(
...
+        }
+        updateViewsContent()
     }
 
-    fun onViewCreated() {
-        if (viewMode == ViewMode.START) {
-            apiManager.sendAction(
-                actionUUID = actionAppLinkData?.actionUUID ?: "",
-                connectionAndKey = richConnection ?: return,
+    private fun sendActionRequest(currentRichConnection: RichConnection, actionID: ID) {
```
[the remark this answers is not in the document]
Comment (ConstantinKV) [reply app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt:102]: it is not correct hierarchy. first should be `public` methods or `public override` methods. The main principle is they should be grouped.

File: authenticator_core/src/main/java/com/saltedge/authenticator/core/model/ActionAppLinkData.kt (hunk 4 above)
Comment (ConstantinKV) [reply authenticator_core/src/main/java/com/saltedge/authenticator/core/model/ActionAppLinkData.kt:36]: in core constants are not available

File: app/src/main/kotlin/com/saltedge/authenticator/features/connections/list/ConnectionsListInteractor.kt (hunk 2 above)
Comment (ConstantinKV) [reply app/src/main/kotlin/com/saltedge/authenticator/features/connections/list/ConnectionsListInteractor.kt:101]: What thing i should choose?

File: app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt (hunk 1 above)
Comment (ConstantinKV) [reply app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt:183]: it tries to open return_to url

Comment (mnewlive) [reply app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt:183]: maybe still open the link?
And in the description of the method to describe what exactly you want to open.

Comment (ConstantinKV) [reply app/src/main/kotlin/com/saltedge/authenticator/features/actions/SubmitActionViewModel.kt:183]: name `openReturnToUrl` is enough

**Pull request merged**